contracts/btc-finality/src/finality.rs

```diff
@@ -588,25 +692,75 @@ pub fn compute_active_finality_providers(
             .into_iter()
             .filter(|fp| {
                 // Filter out FPs with no voting power
-                fp.power > 0
+                if fp.power == 0 {
+                    return false;
+                }
+                // Filter out FPs that are jailed.
+                // Error (shouldn't happen) is being mapped to "jailed forever"
+                JAIL.may_load(deps.storage, &fp.btc_pk_hex)
+                    .unwrap_or(Some(JAIL_FOREVER))
+                    .is_none()
             })
             .scan(total_power, |acc, fp| {
                 *acc += fp.power;
                 Some((fp, *acc))
             })
             .unzip();
+
         finality_providers.extend_from_slice(&filtered);
         total_power = running_total.last().copied().unwrap_or_default();
 
         // and get the next page
         batch = list_fps_by_power(&cfg.staking, &deps.querier, last, QUERY_LIMIT)?;
     }
 
-    // TODO: Online FPs verification (#82)
-    // TODO: Filter out slashed / offline / jailed FPs (#82)
+    // Online FPs verification
+    // Store starting heights of fps entering the active set
+    let old_fps: HashSet<String> = FP_SET
+        .may_load(deps.storage, env.block.height - 1)?
+        .unwrap_or(vec![])
+        .iter()
+        .map(|fp| fp.btc_pk_hex.clone())
+        .collect();
+    let cur_fps: HashSet<String> = finality_providers
+        .iter()
+        .map(|fp| fp.btc_pk_hex.clone())
+        .collect();
+    let new_fps = cur_fps.difference(&old_fps);
+    for fp in new_fps {
+        // Active since the next block. Only save if not already set
+        FP_START_HEIGHT.update(deps.storage, fp, |h| match h {
+            Some(h) => Ok::<_, ContractError>(h),
+            None => Ok(env.block.height + 1),
+        })?;
+    }
+
+    // Check for inactive finality providers, and jail them
+    let params = PARAMS.load(deps.storage)?;
+    finality_providers.iter().try_for_each(|fp| {
+        let mut last_sign_height = FP_BLOCK_SIGNER.may_load(deps.storage, &fp.btc_pk_hex)?;
+        if last_sign_height.is_none() {
+            // Not a block signer yet, check their start height instead
+            last_sign_height = FP_START_HEIGHT.may_load(deps.storage, &fp.btc_pk_hex)?;
+        }
+        match last_sign_height {
+            Some(h) if h > env.block.height.saturating_sub(params.missed_blocks_window) => {
+                Ok::<_, ContractError>(())
+            }
+            _ => {
+                // FP is inactive for at least missed_blocks_window, jail! (if not already jailed)
+                JAIL.update(deps.storage, &fp.btc_pk_hex, |jailed| match jailed {
+                    Some(jail_time) => Ok::<_, ContractError>(jail_time),
+                    None => Ok(env.block.time.seconds() + params.jail_duration),
+                })?;
```

Comment:
Hmm if I understand the design correctly, the jailing condition is as follows: If a FP hasn't submitted `missed_blocks_window` consecutive finality signatures since it becomes active or since its last signature, then the FP will be jailed by the `jail_duration`.

This makes sense to me as a simplified approach, or as a ultimate solution given that we cannot do much heavy-lifting operations in the smart contracts. So maybe we can elaborate this in the ADR

Reply:
This is good enough as a jailing condition for being offline, IMO. The Babylon offline detection is more complex, as it considers a certain percentage of blocks over a given period, or blocks window.

Saying, for example 10% of missed blocks over a 1000 blocks window, is (almost) the same as saying, 1 missed block over a 10 blocks window. The more complex condition can be reduced to a simpler / unitary one, while keeping the same ratio.
I don't think we miss much with this design, and its simplicity is considerably greater.